File: metadata/mill-metadata-core/src/main/java/io/qpointz/mill/metadata/repository/file/FileFacetTypeRepository.java
package io.qpointz.mill.metadata.repository.file;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import io.qpointz.mill.metadata.domain.FacetTypeDescriptor;
import io.qpointz.mill.metadata.repository.FacetTypeRepository;
import lombok.extern.slf4j.Slf4j;

import java.io.IOException;
import java.io.InputStream;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@Slf4j
public class FileFacetTypeRepository implements FacetTypeRepository {

    private final Map<String, FacetTypeDescriptor> store = new ConcurrentHashMap<>();
    private final ObjectMapper yamlMapper;
    private final ResourceResolver resourceResolver;
    private final List<String> locations;

    public FileFacetTypeRepository(List<String> locations, ResourceResolver resourceResolver) {
        this.locations = new ArrayList<>(locations);
        this.resourceResolver = resourceResolver;
        this.yamlMapper = new ObjectMapper(new YAMLFactory());
        this.yamlMapper.registerModule(new JavaTimeModule());
        load();
    }

    private void load() {
        for (String location : locations) {
            try {
                List<ResourceResolver.ResolvedResource> resources = resourceResolver.resolve(location);
                for (ResourceResolver.ResolvedResource resource : resources) {
                    loadFromResource(resource);
                }
            } catch (IOException e) {
                log.warn("Failed to resolve facet type location: {}", location, e);
            }
        }
        log.info("Loaded {} facet type descriptors from {} location(s)", store.size(), locations.size());
    }

    private void loadFromResource(ResourceResolver.ResolvedResource resource) {
        try (InputStream is = resource.inputStream()) {
            FacetTypeFileFormat format = yamlMapper.readValue(is, FacetTypeFileFormat.class);
            if (format.getFacetTypes() != null) {
                for (FacetTypeDescriptor descriptor : format.getFacetTypes()) {
                    store.put(descriptor.getTypeKey(), descriptor);
                    log.debug("Loaded facet type: {} from {}", descriptor.getTypeKey(), resource.name());
                }
            }
        } catch (IOException e) {
            log.warn("Failed to load facet types from: {}", resource.name(), e);
        }
    }

    @Override
    public void save(FacetTypeDescriptor descriptor) {
        store.put(descriptor.getTypeKey(), descriptor);
    }

    @Override
    public Optional<FacetTypeDescriptor> findByTypeKey(String typeKey) {
        return Optional.ofNullable(store.get(typeKey));
    }

    @Override
    public Collection<FacetTypeDescriptor> findAll() {
        return Collections.unmodifiableCollection(store.values());
    }

    @Override
    public void deleteByTypeKey(String typeKey) {
        store.remove(typeKey);
    }

    @Override
    public boolean existsByTypeKey(String typeKey) {
        return store.containsKey(typeKey);
    }

    public static class FacetTypeFileFormat {
        @JsonProperty("facet-types")
        private List<FacetTypeDescriptor> facetTypes;

        public List<FacetTypeDescriptor> getFacetTypes() {
            return facetTypes != null ? facetTypes : List.of();
        }

        public void setFacetTypes(List<FacetTypeDescriptor> facetTypes) {
            this.facetTypes = facetTypes;
        }
    }
}

File: metadata/mill-metadata-core/src/main/kotlin/io/qpointz/mill/metadata/repository/file/FileFacetTypeRepository.kt
package io.qpointz.mill.metadata.repository.file

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import io.qpointz.mill.metadata.domain.FacetTypeDescriptor
import io.qpointz.mill.metadata.repository.FacetTypeRepository
import org.slf4j.LoggerFactory
import java.util.Optional
import java.util.concurrent.ConcurrentHashMap

/** YAML-backed [FacetTypeRepository] that preloads descriptors from configured resources. */
class FileFacetTypeRepository(
    locations: List<String>,
    private val resourceResolver: ResourceResolver
) : FacetTypeRepository {

    private val store = ConcurrentHashMap<String, FacetTypeDescriptor>()
    private val yamlMapper: ObjectMapper = ObjectMapper(YAMLFactory()).apply {
        registerModule(JavaTimeModule())
        registerKotlinModule()
    }

    init {
        for (location in locations) {
            try {
                val resources = resourceResolver.resolve(location)
                for (resource in resources) loadFromResource(resource)
            } catch (e: Exception) {
                log.warn("Failed to resolve facet type location: {}", location, e)
            }
        }
        log.info("Loaded {} facet type descriptors from {} location(s)", store.size, locations.size)
    }

    private fun loadFromResource(resource: ResolvedResource) {
        try {
            resource.inputStream.use { input ->
                val format = yamlMapper.readValue(input, FacetTypeFileFormat::class.java)
                for (descriptor in format.facetTypes) {
                    store[descriptor.typeKey] = descriptor
                    log.debug("Loaded facet type: {} from {}", descriptor.typeKey, resource.name)
                }
            }
        } catch (e: Exception) {
            log.warn("Failed to load facet types from: {}", resource.name, e)
        }
    }

    override fun save(descriptor: FacetTypeDescriptor) {
        store[descriptor.typeKey] = descriptor
    }

    override fun findByTypeKey(typeKey: String): Optional<FacetTypeDescriptor> =
        Optional.ofNullable(store[typeKey])

    override fun findAll(): Collection<FacetTypeDescriptor> =
        store.values.toList()

    override fun deleteByTypeKey(typeKey: String) {
        store.remove(typeKey)
    }

    override fun existsByTypeKey(typeKey: String): Boolean =
        store.containsKey(typeKey)

    /** YAML wrapper object for `facet-types` array. */
    class FacetTypeFileFormat {
        @JsonProperty("facet-types")
        var facetTypes: List<FacetTypeDescriptor> = emptyList()
    }

    companion object {
        private val log = LoggerFactory.getLogger(FileFacetTypeRepository::class.java)
    }
}
